1.<br/>
Polyhedron modeling



(a).<br/>
For the regular cude centered at the orgin,

$$
A = \left[
\begin{array}{ccc} 
0 & 0 & 1\\ 
0 & 1 & 0\\
1 & 0 & 0\\
0 & 0 & -1\\
0 & -1 & 0\\
-1 & 0 & 0
\end{array}\right]
$$

$$
B = \left[
\begin{array}{ccc} 
1\\ 
1\\
1\\
1\\
1\\
1
\end{array}\right]
$$


(b).<br/>
For the Regular octahedron centered at the origin, we have the it described as:

$$
-x_1 + x_2 + x_3 \leq 1\\
x_1 + x_2 + x_3 \leq 1\\
-x_1 - x_2 + x_3 \leq 1\\
x_1 - x_2 + x_3 \leq 1\\
-x_1 + x_2 - x_3 \leq 1\\
-x_1 - x_2 - x_3 \leq 1\\
x_1 + x_2 - x_3 \leq 1\\
x_1 - x_2 - x_3 \leq 1
$$

Therefore,


$$
A = \left[
\begin{array}{ccc} 
-1 & 1 & 1\\ 
1 & 1 & 1\\
-1 & -1 & 1\\
1 & -1 & 1\\
-1 & 1 & -1\\
-1 & -1 & -1\\
1 & 1 & -1\\
1 & -1 & -1
\end{array}\right]
$$

$$
B = \left[
\begin{array}{ccc} 
1\\ 
1\\
1\\
1\\
1\\
1\\
1\\
1
\end{array}\right]
$$

2.<br/>
(a).<br/>
Standard form with equality constraints


we need to turn unbounded variable $z_1$ to bounded by doing:


$z_1 = u - v, u \geq 0, v \geq 0$


we also need to turn bounded variable $z_2, z_3, z_4$ into nonnegative:


$z_2 +1 = l\\
z_3 + 1 = x\\
z_4 + 2 = n$


We also need to add slack variables $s_1, s_2, s_3, s_4, s_5$ to transform inequality to euqality:

$
\begin{aligned}
&(u - v) - 6(l - 1) + (x - 1) - (n - 2) + s_1 = 3\\
&s_1 \geq 0\\
&(x - 1) + (n - 2) + s_2 = 2\\
&s_2 \geq 0\\
&l+s_3 = 6\\
&x+s_4 = 6\\
&n+s_5 = 4\\
&s_3, s_4, s_5 \geq 0
\end{aligned}
$

The optimization problem written in general LP form:

\begin{align*}
\begin{aligned}
& \underset{u, v, l, x, n, s_1, s_2, s_3, s_4, s_5}{\text{minimize}}
& & -3u + 3v + l - 1\\
& \text{subject to} 
\end{aligned} \\
& u - v - 6l + x - n + s_1 = -4\\
& 7l + n = 14\\
& x+n+s_2 = 5\\
& l + s_3 = 6\\
& x + s_4 = 6\\
& n + s_5 = 4\\
& l, x, n, u, v, -s_1, s_2, s_3, s_4, s_5, \geq 0
\end{align*}


Therefore, A, b, c, x are:
$$
A = \left[
\begin{array}{ccc} 
1 & -1 & -6 & 1 & -1 & 1 & 0 & 0 & 0 & 0\\ 
0 & 0 & 7 & 0 & 1 & 0 & 0 & 0 & 0 & 0\\
0 & 0 & 0 & 1 & 1 & 0 & 1 & 0 & 0 & 0\\
0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0\\
0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0\\
0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1
\end{array}\right]
$$
$$
x = \left[
\begin{array}{ccc} 
u\\
v\\
l\\
x\\
n\\
s_1\\
s_2\\
s_3\\
s_4\\
s_5
\end{array}\right]
$$

$$
b = \left[
\begin{array}{ccc} 
-4\\
14\\
5\\
6\\
6\\
4
\end{array}\right]
$$

$$
c = \left[
\begin{array}{ccc} 
-3 & 3 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0
\end{array}\right]
$$


(b).<br/>
Using original version of the LP:

In [30]:
using JuMP, Clp
m = Model(solver = ClpSolver())
@variable(m, z1)
@variable(m, -1 <= z2 <= 5)
@variable(m, -1 <= z3 <= 5)
@variable(m, -2 <= z4 <= 2)
@constraint(m, -z1 + 6z2 - z3 + z4 >= -3)
@constraint(m, 7z2 + z4 >= 5)
@constraint(m, 7z2 + z4 <= 5)
@constraint(m, z3 + z4 <= 2)
@objective(m, Max, 3z1 - z2)

status = solve(m)
println(status)
println("z1 should be : ", getvalue(z1))
println("z2 should be : ", getvalue(z2))
println("z3 should be : ", getvalue(z3))
println("z4 should be : ", getvalue(z4))
println("optimal value is ", getobjectivevalue(m))

Optimal
z1 should be : 8.57142857142857
z2 should be : 0.4285714285714285
z3 should be : -1.0
z4 should be : 2.0
optimal value is 25.28571428571428


Using the transformed version of the LP:


We defined the objective as -3u + 3v + l, therefore the actual optimal cost would be:


-(getobjectivevalue(m)-1)

In [37]:
X = [:u, :v, :l, :x, :n, :s1, :s2, :s3, :s4, :s5]
constraints = [:c1, :c2, :c3, :c4, :c5, :c6]
B = Dict( zip(constraints, [-4, 14, 5, 6, 6, 4]) )
using NamedArrays
A_num = [ 1 1 -6 1 -1 1 0 0 0 0
      0 0 7 0 1 0 0 0 0 0
      0 0 0 1 1 0 1 0 0 0
      0 0 1 0 0 0 0 1 0 0
      0 0 0 1 0 0 0 0 1 0
      0 0 0 0 1 0 0 0 0 1]
A = NamedArray(A_num, (constraints, X), ("constraint", "X"))
C = Dict( zip( X, [ -3, 3, 1, 0, 0, 0, 0, 0, 0, 0 ] ) )

using JuMP, Clp
m = Model(solver=ClpSolver())
@variable(m, val[X] >= 0)
@constraint(m, constr[i in constraints], sum( A[i,s]*val[s] for s in X ) == B[i])
@expression(m, cost, sum( C[s]*val[s] for s in X))
@objective(m, Min, cost)

status = solve(m)

println(status)
println("z1 should be : ", getvalue(val[:u] - val[:v]))
println("z2 should be : ", getvalue(val[:l] - 1))
println("z3 should be : ", getvalue(val[:x] - 1))
println("z4 should be : ", getvalue(val[:n] - 2))
println("optimal value is ", -(getobjectivevalue(m)-1) )

Optimal
z1 should be : 8.571428571428571
z2 should be : 0.4285714285714286
z3 should be : -1.0
z4 should be : 2.0
optimal value is 25.28571428571429


We can see from the above 2 blocks of codes that our two versions give the same result.

3.<br/>
Alloy blending<br/>
For this optimization problem, we have:


Decision variables:<br/>
i1 for iron alloy<br/>
i2 for iron alloy 2<br/>
i3 for iron alloy 3<br/>
c1 for copper 1<br/>
c2 for copper 2<br/>
a1 for aluminum 1<br/>
a2 for aluminum 2<br/>

\begin{aligned}
& \underset{i1, i2, i3, c1, c2, a1, a2}{\text{minimize}}
& & 200i1 + 250i2 + 150i3 + 220c1 + 240c2 + 200a1 + 165a2\\
& \text{subject to} 
\end{aligned}
\begin{aligned}
& 10 \leq 0.025i1 + 0.03i2\leq 15\\
& 2 \leq 0.003i3 + 0.9c1 + 0.96c2 + 0.004a1 + 0.006a2 \leq 3\\
& 6 \leq 0.013i1 + 0.008i2 + 0.04c2 + 0.012a1 \leq 8.25\\
& i1 + i2 + i3 + c1 + c2 + a1 + a2 \geq 500\\
& 0 \leq i1 \leq 400\\
& 0 \leq i2 \leq 300\\
& 0 \leq i3 \leq 600\\
& 0 \leq c1 \leq 500\\
& 0 \leq c2 \leq 200\\
& 0 \leq a1 \leq 300\\
& 0 \leq a2 \leq 250\\
\end{aligned}

In [38]:
using JuMP, Clp
m = Model(solver = ClpSolver())
@variable(m, 0 <= i1 <= 400)
@variable(m, 0 <= i2 <= 300)
@variable(m, 0 <= i3 <= 600)
@variable(m, 0 <= c1 <= 500)
@variable(m, 0 <= c2 <= 200)
@variable(m, 0 <= a1 <= 300)
@variable(m, 0 <= a2 <= 250)
@constraint(m, 10 <= 0.025i1 + 0.03i2 <= 15)
@constraint(m, 2 <= 0.003i3 + 0.9c1 + 0.96c2 + 0.004a1 + 0.006a2 <= 3)
@constraint(m, i1 + i2 + i3 + c1 + c2 + a1 + a2 >= 500)
@constraint(m, 6 <= 0.013i1 + 0.008i2 + 0.04c2 + 0.012a1 <= 8.25)
@objective(m, Min, 200i1 + 250i2 + 150i3 + 220c1 + 240c2 + 200a1 + 165a2)

status = solve(m)

println(status)
println("i1 should be : ", getvalue(i1))
println("i2 should be : ", getvalue(i2))
println("i3 should be : ", getvalue(i3))
println("c1 should be : ", getvalue(c1))
println("c2 should be : ", getvalue(c2))
println("a1 should be : ", getvalue(a1))
println("a2 should be : ", getvalue(a2))
println("optimal value is ", getobjectivevalue(m))

Optimal
i1 should be : 400.0
i2 should be : 0.0
i3 should be : 39.77630199231039
c1 should be : 0.0
c2 should be : 2.761272282418735
a1 should be : 57.462425725270876
a2 should be : 0.0
optimal value is 98121.63579168124


Therefore, we should use 400 tons of iron alloy 1, 39.77630199231039 tons of iron alloy 2, 2.761272282418735 tons of copper 2, 57.462425725270876 tons of aluminum 1.

4.<br/>
Stigler’s diet<br/>
(a).<br/>

In [39]:
# STARTER CODE FOR STIGLER'S DIET PROBLEM
using NamedArrays

# import Stigler's data set
raw = readcsv("stigler.csv")
(m,n) = size(raw)

n_nutrients = 2:n    # columns containing nutrients
n_foods = 3:m    # rows containing food names

nutrients = raw[1,n_nutrients][:]    # the list of nutrients (convert to 1-D array)
foods = raw[n_foods,1][:]    # the list of foods (convert to 1-D array)

# lower[i] is the minimum daily requirement of nutrient i.
lower = Dict( zip(nutrients,raw[2,n_nutrients]) )

# data[f,i] is the amount of nutrient i contained in food f.
data = NamedArray( raw[n_foods,n_nutrients], (foods,nutrients), ("foods","nutrients") );

In [40]:
using JuMP, Clp
m = Model(solver = ClpSolver())

@variable(m, diet[foods] >= 0)
println(diet)
@constraint(m, constr[i in nutrients], sum( data[s,i]*diet[s] for s in foods ) >= lower[i])
@expression(m, total_cost, sum(diet[s] for s in foods))
@objective(m, Min, total_cost)

status = solve(m)
println(status)
println(getvalue(diet))
println("Total cost is: \$", getvalue(total_cost))

diet[i] ≥ 0 ∀ i ∈ {Wheat Flour (Enriched),Macaroni,…,Molasses,Strawberry Preserves}
Optimal
diet: 1 dimensions:
[ Wheat Flour (Enriched)] = 0.02951906167648827
[               Macaroni] = 0.0
[Wheat Cereal (Enriched)] = 0.0
[            Corn Flakes] = 0.0
[              Corn Meal] = 0.0
[           Hominy Grits] = 0.0
[                   Rice] = 0.0
[            Rolled Oats] = 0.0
[ White Bread (Enriched)] = 0.0
[      Whole Wheat Bread] = 0.0
[              Rye Bread] = 0.0
[             Pound Cake] = 0.0
[          Soda Crackers] = 0.0
[                   Milk] = 0.0
[  Evaporated Milk (can)] = 0.0
[                 Butter] = 0.0
[          Oleomargarine] = 0.0
[                   Eggs] = 0.0
[       Cheese (Cheddar)] = 0.0
[                  Cream] = 0.0
[          Peanut Butter] = 0.0
[             Mayonnaise] = 0.0
[                 Crisco] = 0.0
[                   Lard] = 0.0
[          Sirloin Steak] = 0.0
[            Round Steak] = 0.0
[              Rib Roast] = 0.0
[       

According to the result given by the linear solver, the cost for a year is $0.10866227820675685 \times 365 = 39.6617$ dollors. This answer is pretty close to the answer given by Stigler(39.93 dollors).


The foods on the list are: Wheat Flour (Enriched), Liver (Beef), Cabbage, Spinach, Navy Beans, Dried

(b).<br/>
The cheapest diet for a vegetarian could be given by the same optimization with non-vegetarian items removed from the .csv file:

In [41]:
# STARTER CODE FOR STIGLER'S DIET PROBLEM
using NamedArrays

# import Stigler's data set
raw = readcsv("stigler_vegetarian.csv")
(m,n) = size(raw)

n_nutrients = 2:n    # columns containing nutrients
n_foods = 3:m    # rows containing food names

nutrients = raw[1,n_nutrients][:]    # the list of nutrients (convert to 1-D array)
foods = raw[n_foods,1][:]    # the list of foods (convert to 1-D array)

# lower[i] is the minimum daily requirement of nutrient i.
lower = Dict( zip(nutrients,raw[2,n_nutrients]) )

# data[f,i] is the amount of nutrient i contained in food f.
data = NamedArray( raw[n_foods,n_nutrients], (foods,nutrients), ("foods","nutrients") );

In [42]:
using JuMP, Clp
m = Model(solver = ClpSolver())

@variable(m, diet[foods] >= 0)
println(diet)
@constraint(m, constr[i in nutrients], sum( data[s,i]*diet[s] for s in foods ) >= lower[i])
@expression(m, total_cost, sum(diet[s] for s in foods))
@objective(m, Min, total_cost)

status = solve(m)
println(status)
println(getvalue(diet))
println("Total cost is: \$", getvalue(total_cost))

diet[i] ≥ 0 ∀ i ∈ {Wheat Flour (Enriched),Macaroni,…,Molasses,Strawberry Preserves}
Optimal
diet: 1 dimensions:
[ Wheat Flour (Enriched)] = 0.035455581408887715
[               Macaroni] = 0.0
[Wheat Cereal (Enriched)] = 0.0
[            Corn Flakes] = 0.0
[              Corn Meal] = 0.0
[           Hominy Grits] = 0.0
[                   Rice] = 0.0
[            Rolled Oats] = 0.0
[ White Bread (Enriched)] = 0.0
[      Whole Wheat Bread] = 0.0
[              Rye Bread] = 0.0
[             Pound Cake] = 0.0
[          Soda Crackers] = 0.0
[                   Milk] = 0.0
[  Evaporated Milk (can)] = 0.008591461668763569
[                 Butter] = 0.0
[          Oleomargarine] = 0.0
[                   Eggs] = 0.0
[       Cheese (Cheddar)] = 0.0
[                  Cream] = 0.0
[          Peanut Butter] = 0.0
[             Mayonnaise] = 0.0
[                 Crisco] = 0.0
[                   Lard] = 0.0
[                  Plate] = 0.0
[                 Apples] = 0.0
[                Banan

In this case, the total cost for a day is $0.10903743657646292 dollors, therefore the cost for a year is 39.7987 dollors. We will use Wheat Flour (Enriched), Evaporated Milk (can), Cabbage, Spinach, Navy Beans, Dried.